### PR TITLE
[k8s] kube-client: refactor error handling

### PR DESCRIPTION
**Pull request by dmolokanov**
Refactor error handling for kube-client crate

### CODE REVIEW

File: edgelet/kube-client/src/error.rs [hunk 1]
```diff
@@ -103,113 +87,70 @@ impl From<Context<ErrorKind>> for Error {
...
+            }
+            Self::MissingOrInvalidKubeContext => write!(
+                f,
+                "Missing or invalid Kubernetes context in .kube/config file."
+            ),
+            Self::MissingUser => write!(f, "Missing user configuration in .kube/config file."),
+            Self::MissingData => write!(f, "Both file and data missing."),
+            Self::Base64Decode => write!(f, "Base64 decode error."),
+            Self::UrlParse(x) => write!(f, "Unable to parse valid URL from: {}.", x),
+            Self::Tls => write!(f, "Could create TLS connector."),
```
Comment (arsing) [inline edgelet/kube-client/src/error.rs]: `Could not`

File: edgelet/iotedge-proxy/src/error.rs [hunk 2]
```diff
@@ -54,23 +54,19 @@ pub enum InitializeErrorReason {
...
         match self {
-            InitializeErrorReason::LoadSettings => write!(f, "Could not load settings"),
-            InitializeErrorReason::LoadSettingsUnsupportedSchema(x) => {
-                write!(f, "Unsupported URL schema: {}", x)
-            }
-            InitializeErrorReason::InvalidUrl(x) => {
-                write!(f, "Could not resolve address for given URL: {}", x)
-            }
-            InitializeErrorReason::InvalidUrlWithReason(url, reason) => write!(
+            Self::LoadSettings => write!(f, "Could not load settings"),
```
Comment (darobs) [inline edgelet/iotedge-proxy/src/error.rs]: That looks much better.

File: edgelet/kube-client/src/client.rs [hunk 3]
```diff
@@ -135,18 +147,21 @@ where
             config_map,
             api_core::ReplaceNamespacedConfigMapOptional::default(),
         )
-        .map_err(Error::from)
+        .map_err(|err| Error::from(err.context(ErrorKind::Request(RequestType::ConfigMapCreate))))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapCreate [](start = 79, length = 15)

ConfigMapReplace

File: edgelet/kube-client/src/client.rs [hunk 4]
```diff
@@ -110,14 +116,20 @@ where
...
+                .and_then(|response| match response {
+                    api_core::CreateNamespacedConfigMapResponse::Accepted(config_map)
+                    | api_core::CreateNamespacedConfigMapResponse::Created(config_map)
+                    | api_core::CreateNamespacedConfigMapResponse::Ok(config_map) => Ok(config_map),
+                    _ => Err(Error::from(ErrorKind::Response(
+                        RequestType::ConfigMapCreate,
+                    ))),
+                })
+                .map_err(|err| {
+                    Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 77, length = 13)

ConfigMapCreate

File: edgelet/kube-client/src/client.rs [hunk 5]
```diff
@@ -135,18 +147,21 @@ where
...
+                    api_core::ReplaceNamespacedConfigMapResponse::Ok(config_map)
+                    | api_core::ReplaceNamespacedConfigMapResponse::Created(config_map) => {
+                        Ok(config_map)
+                    }
+                    _ => Err(Error::from(ErrorKind::Response(
+                        RequestType::ConfigMapReplace,
+                    ))),
+                })
+                .map_err(|err| {
+                    Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 77, length = 13)

ConfigMapReplace

File: edgelet/kube-client/src/client.rs [hunk 6]
```diff
@@ -162,13 +177,19 @@ where
...
+            self.request(req)
+                .and_then(|response| match response {
+                    api_core::DeleteNamespacedConfigMapResponse::OkStatus(_)
+                    | api_core::DeleteNamespacedConfigMapResponse::OkValue(_) => Ok(()),
+                    _ => Err(Error::from(ErrorKind::Response(
+                        RequestType::ConfigMapDelete,
+                    ))),
+                })
+                .map_err(|err| {
+                    Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 77, length = 13)

ConfigMapDelete

File: edgelet/kube-client/src/client.rs [hunk 7]
```diff
@@ -188,12 +209,22 @@ where
...
+                    .and_then(|response| match response {
+                        api_apps::ListNamespacedDeploymentResponse::Ok(deployments) => {
+                            Ok(deployments)
+                        }
+                        _ => Err(Error::from(ErrorKind::Response(
+                            RequestType::DeploymentList,
+                        ))),
+                    })
+                    .map_err(|err| {
+                        Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 81, length = 13)

DeploymentList

File: edgelet/kube-client/src/client.rs [hunk 8]
```diff
@@ -209,14 +240,22 @@ where
...
+                    | api_apps::CreateNamespacedDeploymentResponse::Created(deployment)
+                    | api_apps::CreateNamespacedDeploymentResponse::Ok(deployment) => {
+                        Ok(deployment)
+                    }
+                    _ => Err(Error::from(ErrorKind::Response(
+                        RequestType::DeploymentCreate,
+                    ))),
+                })
+                .map_err(|err| {
+                    Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 77, length = 13)

DeploymentCreate

File: edgelet/kube-client/src/client.rs [hunk 9]
```diff
@@ -234,13 +273,21 @@ where
...
+                    api_apps::ReplaceNamespacedDeploymentResponse::Created(deployment)
+                    | api_apps::ReplaceNamespacedDeploymentResponse::Ok(deployment) => {
+                        Ok(deployment)
+                    }
+                    _ => Err(Error::from(ErrorKind::Response(
+                        RequestType::DeploymentReplace,
+                    ))),
+                })
+                .map_err(|err| {
+                    Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 77, length = 13)

DeploymentReplace

File: edgelet/kube-client/src/client.rs [hunk 10]
```diff
@@ -256,12 +303,16 @@ where
...
-                    api_core::ListNamespacedPodResponse::Ok(list) => Ok(list),
-                    _ => Err(Error::from(ErrorKind::Response)),
-                })
+                self.request(req)
+                    .and_then(|response| match response {
+                        api_core::ListNamespacedPodResponse::Ok(list) => Ok(list),
+                        _ => Err(Error::from(ErrorKind::Response(RequestType::PodList))),
+                    })
+                    .map_err(|err| {
+                        Error::from(err.context(ErrorKind::Response(RequestType::ConfigMapList)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: >ConfigMapList [](start = 81, length = 13)

PodList

File: edgelet/kube-client/src/lib.rs [hunk 11]
```diff
@@ -16,4 +16,4 @@ pub mod kube;
 
 pub use self::client::{Client, HttpClient};
 pub use self::config::{get_config, Config, TokenSource, ValueToken};
-pub use self::error::{Error, ErrorKind};
+pub use self::error::{Error, ErrorKind, KubeConfigErrorReason, RequestType};
```
Comment (darobs) [inline edgelet/kube-client/src/lib.rs]: >KubeConfigErrorReason, RequestType [](start = 40, length = 34)

Does this need to be public?

File: edgelet/kube-client/src/client.rs [hunk 12]
```diff
@@ -453,15 +544,25 @@ where
             service_account,
             api_core::ReplaceNamespacedServiceAccountOptional::default(),
         )
-        .map_err(Error::from)
+        .map_err(|err| {
+            Error::from(err.context(ErrorKind::Request(RequestType::ServiceAccountCreate)))
```
Comment (darobs) [inline edgelet/kube-client/src/client.rs]: ServiceAccountReplace

**Pull request merged**